File: ios-app/IzmirEshot/Networking/EshotService.swift
import Foundation

enum APIError: Error {
    case invalidURL
    case noData
    case decodingError
}

class EshotService {
    static let shared = EshotService()
    
    private init() {}
    
    func fetchApproachingBuses(stopId: String) async throws -> [BusDto] {
        guard let url = URL(string: "https://openapi.izmir.bel.tr/api/iztek/duragayaklasanotobusler/\(stopId)") else {
            throw APIError.invalidURL
        }
        
        let (data, _) = try await URLSession.shared.data(from: url)
        
        let decoder = JSONDecoder()
        // Lenient decoding handled in DTO properties (String -> Double)
        return try decoder.decode([BusDto].self, from: data)
    }
    
    // Using CSV for Nearby finding logic as per Plan (reliable WGS84)
    // But if we want to use the API:
    func fetchNearbyStops(lat: Double, lon: Double) async throws -> Data {
        guard let url = URL(string: "https://openapi.izmir.bel.tr/api/ibb/cbs/noktayayakinduraklar?x=\(lon)&y=\(lat)&inCoordSys=WGS84&outCoordSys=WGS84") else {
             throw APIError.invalidURL
        }
        let (data, _) = try await URLSession.shared.data(from: url)
        return data
    }
}

File: android-app/app/src/main/java/com/eshot/app/data/api/EshotService.kt
package com.eshot.app.data.api

import com.eshot.app.data.model.BusDto
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.android.Android
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.json.Json

object EshotService {
    private val client = HttpClient(Android) {
        install(ContentNegotiation) {
            json(Json {
                ignoreUnknownKeys = true
                isLenient = true
            })
        }
    }

    suspend fun getApproachingBuses(stopId: String): List<BusDto> {
        return try {
            client.get("https://openapi.izmir.bel.tr/api/iztek/duragayaklasanotobusler/$stopId").body()
        } catch (e: Exception) {
            emptyList()
        }
    }
    
    suspend fun getNearbyStopsRaw(lat: Double, lon: Double): String {
       return client.get("https://openapi.izmir.bel.tr/api/ibb/cbs/noktayayakinduraklar?x=$lon&y=$lat&inCoordSys=WGS84&outCoordSys=WGS84").body()
    }
}
